File: easy/ts/112.ts
// https://leetcode.com/problems/path-sum/
// Time and Space: O(n)

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
  return checkPathSum(root, 0, targetSum);
}

function checkPathSum(
  node: TreeNode | null,
  pathSum: number,
  targetSum: number
): boolean {
  if (!node) return false;

  pathSum += node.val;

  // If we are at a leaf node and the path sum matches the target
  if (!node.left && !node.right && pathSum === targetSum) {
    return true;
  }

  return (
    checkPathSum(node.left, pathSum, targetSum) ||
    checkPathSum(node.right, pathSum, targetSum)
  );
}

File: easy/python/112.py
# https://leetcode.com/problems/path-sum/description/
# Time and Space: O(n)
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False

        stack = [(root, root.val)]

        while stack:
            node, current_sum = stack.pop()

            if not node.left and not node.right and current_sum == targetSum:
                return True

            if node.right:
                stack.append((node.right, current_sum + node.right.val))

            if node.left:
                stack.append((node.left, current_sum + node.left.val))

        return False
